src/time_integration/paired_explicit_runge_kutta/methods_PERK3.jl

```diff
@@ -0,0 +1,366 @@
+# which are used in Trixi.
+mutable struct PairedExplicitRK3Integrator{RealT <: Real, uType, Params, Sol, F, Alg,
+                                           PairedExplicitRKOptions} <:
+               AbstractPairedExplicitRKSingleIntegrator
+    u::uType
+    du::uType
+    u_tmp::uType
+    t::RealT
+    tdir::RealT
+    dt::RealT # current time step
+    dtcache::RealT # manually set time step
+    iter::Int # current number of time steps (iteration)
+    p::Params # will be the semidiscretization from Trixi
+    sol::Sol # faked
+    f::F
+    alg::Alg # This is our own class written above; Abbreviation for ALGorithm
+    opts::PairedExplicitRKOptions
+    finalstep::Bool # added for convenience
+    dtchangeable::Bool
+    force_stepfail::Bool
+    # PairedExplicitRK stages:
+    k1::uType
+    k_higher::uType
+end
+
+function init(ode::ODEProblem, alg::PairedExplicitRK3;
+              dt, callback = nothing, kwargs...)
+    u0 = copy(ode.u0)
+    du = zero(u0)
+    u_tmp = zero(u0)
+
+    # PairedExplicitRK stages
+    k1 = zero(u0)
+    k_higher = zero(u0)
+
+    t0 = first(ode.tspan)
+    tdir = sign(ode.tspan[end] - ode.tspan[1])
+    iter = 0
+
+    integrator = PairedExplicitRK3Integrator(u0, du, u_tmp, t0, tdir, dt, dt, iter,
+                                             ode.p,
+                                             (prob = ode,), ode.f, alg,
+                                             PairedExplicitRKOptions(callback,
+                                                                     ode.tspan;
+                                                                     kwargs...),
+                                             false, true, false,
+                                             k1, k_higher)
+
+    # initialize callbacks
+    if callback isa CallbackSet
+        for cb in callback.continuous_callbacks
+            error("unsupported")
+        end
+        for cb in callback.discrete_callbacks
+            cb.initialize(cb, integrator.u, integrator.t, integrator)
+        end
+    elseif !isnothing(callback)
+        error("unsupported")
+    end
+
+    return integrator
+end
+
+# Fakes `solve`: https://diffeq.sciml.ai/v6.8/basics/overview/#Solving-the-Problems-1
+function solve(ode::ODEProblem, alg::PairedExplicitRK3;
+               dt, callback = nothing, kwargs...)
+    integrator = init(ode, alg, dt = dt, callback = callback; kwargs...)
+
+    # Start actual solve
+    solve!(integrator)
+end
+
+function solve!(integrator::PairedExplicitRK3Integrator)
+    @unpack prob = integrator.sol
+
+    integrator.finalstep = false
+
+    @trixi_timeit timer() "main loop" while !integrator.finalstep
+        step!(integrator)
+    end # "main loop" timer
+
+    return TimeIntegratorSolution((first(prob.tspan), integrator.t),
+                                  (prob.u0, integrator.u),
+                                  integrator.sol.prob)
+end
+
+function step!(integrator::PairedExplicitRK3Integrator)
```

Comment:
Some questions regarding the interface from the SciML ecosystem (I'm not really familiar with it, so these questions could also be stupid):
1. How does this `step!` function compare to what is usually called `perform_step!` in OrdinaryDiffEq.jl? Is it the same (then why don't call it `perform_step!` for consistency)? If not, what are the differences?
2. In OrdinaryDiffEq.jl we obtain some statistics from `integrator.stats`, e.g., `integrator.stats.nf` can be quite useful sometimes. We don't get this here, do we? And if we don't get it, would it be too hard to add this functionality? Would be nice to have, also in view of being consistent with the OrdinaryDiffEq.jl solvers.
3. Does this work with error-based adaptive time stepping out of the box or do we need either a fixed time step or (as you use in the elixir) a `StepsizeCallback`?

Reply:
I too am not very familiar with SciML so I hope that these, more or less, answer your questions logically.

1. I think that generally, step! here and perform_step! fundamentally do the same thing. But with a little more handling of scenarios in our `step!`, since `step!` implemented here is supposed to be an attempt to replicate the functionality of [step!](https://github.com/SciML/OrdinaryDiffEq.jl/blob/master/src/iterator_interface.jl) in `OrdinaryDiffEq.jl`.

2. I think It's possible to implement the suggested change, but this would require similar modifications across other integrators with the same structure.

3. This approach is not compatible with error-based adaptive time stepping; it only works with a fixed time step or with StepsizeCallback.

Our implementation of step!, solve!, and init! addresses the concerns raised in issue #1886. The refactoring in [#1886](https://github.com/trixi-framework/Trixi.jl/issues/1886), applied to some integrators in Trixi.jl, might not fully align with OrdinaryDiffEq.jl, and there could be room for improvement. But IMHO, it might be a good idea to create another issue or reopen the old one regarding this. 

@ranocha, what do you think about this?

Reply:
Thanks for the explanation.
1. I see that the `step!` function you quote calls `perform_step!`. Would it make sense to have a similar structure here, too, i.e. having both a `step!` and a `perform_step!`, where the latter is called from the former.

But, as you said, this doesn't need to be implemented in this PR and could maybe be simply tracked by #1886 after reopening it? I think it would be nice to have an even more similar structure than in OrdinaryDiffEq.jl. But this should also be done for all our custom ODE solvers uniformly.

Reply:
I guess one could put in a lot more effort to make our simple integrators more similar to what is there in the SciML code. Note, however, that the motivation for #1886 comes from a certain use case which only requires that we have `step!` and `init`. 

Looking for instance at `methods_2N.jl` one could certainly extract

https://github.com/trixi-framework/Trixi.jl/blob/e1aabb309249841e78d42da5072dde21989a9f11/src/time_integration/methods_2N.jl#L175-L190

into something like `perform_step!`. 
As this gives no real needed new functionality, I would consider this a low priority. Still, we could open an issue noting this @JoshuaLampert for the future.